### PR TITLE
[WIP] feat: introduce `foundry-compilers`

### PR DESCRIPTION
**Pull request by Evalir**
Introduces foundry-compilers to the codebase. This will lead to us removing the remaining ABI and primitive types from all the affected crates.

### CODE REVIEW

File: crates/evm/src/trace/decoder.rs [hunk 1]
```diff
@@ -144,25 +140,25 @@ impl CallTraceDecoder {
             // TODO: These are the Ethereum precompiles. We should add a way to support precompiles
             // for other networks, too.
             precompiles: precompiles!(
-                0x01: ecrecover(hash: FixedBytes(32), v: Uint(256), r: Uint(256), s: Uint(256)) -> (publicAddress: Address),
```
Comment (gakonst) [inline crates/evm/src/trace/decoder.rs:146]: This change looks a bit odd cc @DaniPopes

File: crates/evm/src/trace/mod.rs [hunk 2]
```diff
@@ -2,11 +2,8 @@ use crate::{
     abi::CHEATCODE_ADDRESS, debug::Instruction, trace::identifier::LocalTraceIdentifier, CallKind,
 };
 pub use decoder::{CallTraceDecoder, CallTraceDecoderBuilder};
-use ethers::{
-    abi::{ethereum_types::BigEndianHash, Address, RawLog},
-    core::utils::to_checksum,
-    types::{Bytes, DefaultFrame, GethDebugTracingOptions, StructLog, H256, U256},
-};
+use ethers::types::{DefaultFrame, GethDebugTracingOptions, StructLog};
```
Comment (gakonst) [inline crates/evm/src/trace/mod.rs:5]: These types should also go to alloy? I guess RPC crate under core/?

Think good opportunity to get that started, along with the TxRequest/Response/Receipt?

File: crates/evm/src/trace/mod.rs [hunk 3]
```diff
@@ -543,7 +540,7 @@ impl fmt::Display for CallTrace {
                 self.gas_cost,
                 color.paint(self.label.as_ref().unwrap_or(&address)),
                 color.paint(func),
-                if !self.value.is_zero() {
+                if !self.value == U256::ZERO {
```
Comment (gakonst) [inline crates/evm/src/trace/mod.rs:543]: We should impl that on u256

File: crates/evm/src/trace/node.rs [hunk 4]
```diff
@@ -110,19 +109,19 @@ impl CallTraceNode {
 
         if let RawOrDecodedCall::Raw(ref bytes) = self.trace.data {
             let inputs = if bytes.len() >= SELECTOR_LEN {
-                if self.trace.address == CHEATCODE_ADDRESS.to_ethers() {
+                if self.trace.address == CHEATCODE_ADDRESS {
                     // Try to decode cheatcode inputs in a more custom way
                     utils::decode_cheatcode_inputs(func, bytes, errors, verbosity).unwrap_or_else(
                         || {
-                            func.decode_input(&bytes[SELECTOR_LEN..])
+                            func.abi_decode_input(&bytes[SELECTOR_LEN..], false)
```
Comment (gakonst) [inline crates/evm/src/trace/node.rs:116]: @DaniPopes should this be true?

**Review by gakonst — COMMENTED** [review]
Good start

File: crates/evm/src/trace/mod.rs (hunk 3 above)
Comment (DaniPopes) [reply crates/evm/src/trace/mod.rs:543]: it's on ruint but not released yet iirc

File: crates/evm/src/fuzz/strategies/int.rs [hunk 5]
```diff
@@ -159,7 +150,6 @@ impl IntStrategy {
             x if (128..256).contains(&x) => higher &= (1u128 << (x - 128)) - 1,
             _ => {}
         };
-
```
Comment (DaniPopes) [inline crates/evm/src/fuzz/strategies/int.rs:162]: What's with the whitespace deletions?

File: crates/evm/src/fuzz/strategies/param.rs [hunk 6]
```diff
@@ -1,54 +1,55 @@
...
 use proptest::prelude::*;
 
 /// The max length of arrays we fuzz for is 256.
 pub const MAX_ARRAY_LEN: usize = 256;
-
 /// Given a parameter type, returns a strategy for generating values for that type.
 ///
 /// Works with ABI Encoder v2 tuples.
-pub fn fuzz_param(param: &ParamType) -> BoxedStrategy<Token> {
+pub fn fuzz_param(param: &DynSolType) -> BoxedStrategy<DynSolValue> {
```
Comment (DaniPopes) [inline crates/evm/src/fuzz/strategies/param.rs:11]: both value and type implement Proptest traits, you should use those

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/evm/src/fuzz/strategies/state.rs [hunk 7]
```diff
@@ -101,9 +95,9 @@ pub fn build_initial_state<DB: DatabaseRef>(
     let mut state = FuzzDictionary::default();
 
     for (address, account) in db.accounts.iter() {
-        let address: Address = address.to_ethers();
+        let address: Address = *address;
         // Insert basic account information
-        state.values_mut().insert(H256::from(address).into());
+        state.values_mut().insert(B256::from_slice(address.as_slice()).into());
```
Comment (DaniPopes) [inline crates/evm/src/fuzz/strategies/state.rs:100]: This panics. Use `address.into_word()`. Same for all other `B256::from_slice`

File: crates/evm/src/fuzz/strategies/uint.rs [hunk 8]
```diff
@@ -101,42 +93,34 @@ impl UintStrategy {
...
-
     fn generate_edge_tree(&self, runner: &mut TestRunner) -> NewTree<Self> {
         let rng = runner.rng();
-
         // Choose if we want values around 0 or max
         let is_min = rng.gen_bool(0.5);
         let offset = U256::from(rng.gen_range(0..4));
         let max = if self.bits < 256 {
-            (U256::from(1u8) << U256::from(self.bits)) - 1
+            (U256::from(1u8).rotate_left(self.bits)) - U256::from(1)
```
Comment (DaniPopes) [inline crates/evm/src/fuzz/strategies/uint.rs:102]: rotate is not the same as shift

File: crates/evm/src/trace/node.rs (hunk 4 above)
Comment (DaniPopes) [reply crates/evm/src/trace/node.rs:116]: `false` has the same behavior as `ethabi`. Doesn't really matter here

File: crates/evm/src/trace/decoder.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/evm/src/trace/decoder.rs:146]: The whole point of this macro is to make it easy to modify. Slapping `format` everywhere doesn't really help towards that. Anyway I am working on removing this in #5998 since some precompiles don't adhere to ABI so it doesn't matter.

File: crates/evm/src/fuzz/strategies/param.rs [hunk 9]
```diff
@@ -60,90 +31,101 @@ pub fn fuzz_param_from_state(param: &ParamType, arc_state: EvmFuzzState) -> Boxe
...
-            (60, fuzz_calldata(func.clone())),
-            (40, fuzz_calldata_from_state(func, state)),
-        ]);
-
-        let cfg = proptest::test_runner::Config { failure_persistence: None, ..Default::default() };
-        let mut runner = proptest::test_runner::TestRunner::new(cfg);
-
-        let _ = runner.run(&strat, |_| Ok(()));
-    }
+    // TODO: Need a human readable function parser to re-enable the test.
```
Comment (DaniPopes) [inline crates/evm/src/fuzz/strategies/param.rs:116]: `Function::parse`

File: crates/evm/src/fuzz/strategies/param.rs [hunk 10]
```diff
@@ -60,90 +31,101 @@ pub fn fuzz_param_from_state(param: &ParamType, arc_state: EvmFuzzState) -> Boxe
...
         }
-        ParamType::Tuple(params) => params
+        DynSolType::Tuple(params) => params
             .iter()
             .map(|p| fuzz_param_from_state(p, arc_state.clone()))
             .collect::<Vec<_>>()
-            .prop_map(Token::Tuple)
+            .prop_map(DynSolValue::Tuple)
             .boxed(),
+        _ => panic!("Unimplemented"),
```
Comment (DaniPopes) [inline crates/evm/src/fuzz/strategies/param.rs:104]: `unimplemented!()`
This should be unreachable if eip712 feature is disabled

File: crates/abi/src/bindings/alloy_console.rs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+//! Alloy console bindings.
```
Comment (DaniPopes) [inline crates/abi/src/bindings/alloy_console.rs:1]: What is this for? I wouldn't do this to avoid conflicts with #5998

File: crates/abi/src/bindings/mod.rs [hunk 12]
```diff
@@ -6,3 +6,4 @@
 pub mod console;
 pub mod hardhat_console;
 pub mod hevm;
+pub mod alloy_console;
```
Comment (DaniPopes) [inline crates/abi/src/bindings/mod.rs:9]: This is not generated by abigen

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/abi/src/bindings/alloy_console.rs (hunk 11 above)
Comment (Evalir) [reply crates/abi/src/bindings/alloy_console.rs:1]: This is for migrating the console bindings so decoding utilities can be migrated—but happy to do it after #5998, as it's closely related to cheatcodes

File: crates/evm/src/executor/mod.rs [hunk 13]
```diff
@@ -842,10 +843,16 @@ fn convert_call_result<D: Detokenize>(
 
     match status {
         return_ok!() => {
-            let result = decode_function_data(func, result, false)?;
+            let mut result = func.abi_decode_output(&result, false)?;
```
Comment (Evalir) [inline crates/evm/src/executor/mod.rs:846]: I'm not sure using the dynamic encoder with the executor is the right way to approach this—as disambiguating with `DynSolValue` then becomes a tad cumbersome when using them. Do you think there's a way to use the static encoder here maybe with `SolValue` @DaniPopes ?

File: crates/evm/src/decode.rs [hunk 14]
```diff
@@ -26,52 +55,65 @@ pub fn decode_console_logs(logs: &[Log]) -> Vec<String> {
...
-        LogNamedArray1Filter(inner) => format!("{}: {:?}", inner.key, inner.val),
-        LogNamedArray2Filter(inner) => format!("{}: {:?}", inner.key, inner.val),
-        LogNamedArray3Filter(inner) => format!("{}: {:?}", inner.key, inner.val),
+    let raw_log = AlloyLog::new_unchecked(log.topics.into_iter().map(|h| h.to_alloy()).collect_vec(), log.data.0.into());
+    let data = log_address::abi_decode_data(&raw_log.data, false);
+    if let Ok(inner) = log::abi_decode_data(&raw_log.data, false) {
+        return Some(inner.0)
+    } else if let Ok(inner) = logs::abi_decode_data(&raw_log.data, false) {
+        return Some(format!("{}", Bytes::from(inner.0)))
+    } else if let Ok(inner) = log_address::abi_decode_data(&raw_log.data, false) {
```
Comment (gakonst) [inline crates/evm/src/decode.rs]: cc @DaniPopes there should be a better way to do this?

File: crates/evm/src/fuzz/strategies/param.rs [hunk 15]
```diff
@@ -1,149 +1,158 @@
...
+            proptest::collection::vec(fuzz_param(param), fixed_size)
+                .prop_map(DynSolValue::FixedArray)
+                .sboxed()
         }
+        DynSolType::Array(param) => proptest::collection::vec(fuzz_param(param), 0..MAX_ARRAY_LEN)
+            .prop_map(DynSolValue::Array)
+            .sboxed(),
+        DynSolType::CustomStruct { .. } => panic!("unsupported type"),
     }
 }
```
Comment (Evalir) [inline crates/evm/src/fuzz/strategies/param.rs:47]: Wondering what's going on here—fuzzing failures in CI are due to this fn generating invalid values (e.g `uint256`s for `uint8`s and so on), making the function to fuzz revert instantly. Do these need to be constrained further? cc @DaniPopes

File: crates/evm/src/decode.rs [hunk 16]
```diff
@@ -26,52 +49,149 @@ pub fn decode_console_logs(logs: &[Log]) -> Vec<String> {
...
+    );
+    if let Ok(inner) =
+        log::decode_log(raw_log.topics().iter().map(|t| WordToken::from(*t)), &raw_log.data, true)
+    {
+        return Some(inner._0)
+    } else if let Ok(inner) =
+        logs::decode_log(raw_log.topics().iter().map(|t| WordToken::from(*t)), &raw_log.data, true)
+    {
+        return Some(format!("{}", Bytes::from(inner._0)))
+    } else if let Ok(inner) = log_address::decode_log(
```
Comment (Evalir) [inline crates/evm/src/decode.rs:64]: This seems to also be decoding `uint256`s if they fit—this shouldn't be possible? e.g see this test failure on CI run: https://github.com/foundry-rs/foundry/actions/runs/6612838157/job/17960206523?pr=5986#step:9:197

**Review by Evalir — COMMENTED** [review]
This already compiles and runs—getting CI green right now. Some observations on what's broken (fuzzer):

File: crates/evm/src/decode.rs (hunk 16 above)
Comment (Evalir) [reply crates/evm/src/decode.rs]: nvm—this functionality is not yet in Alloy, so reverting to abigen for now.

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/evm/src/executor/abi/mod.rs [hunk 17]
```diff
@@ -38,6 +39,641 @@ pub fn patch_hardhat_console_selector(input: &mut Vec<u8>) {
     }
 }
 
+pub static HEVM_ABI: Lazy<JsonAbi> = Lazy::new(|| {
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/evm/src/executor/abi/mod.rs:42]: This is used on the CallTraceDecoder though, so we need them here with the alloy type—or are you thinking something different?

File: crates/cast/bin/cmd/create2.rs [hunk 18]
```diff
@@ -124,14 +120,9 @@ impl Create2Args {
...
-                let salt = H256::random_using(&mut thread_rng());
-                let salt = Bytes::from(salt.to_fixed_bytes());
+                let salt = B256::random();
 
-                let addr = SimpleCast::to_checksum_address(&get_create2_address_from_hash(
-                    deployer,
-                    salt.clone(),
-                    init_code_hash,
-                ));
+                let addr = SimpleCast::to_checksum_address(&deployer.create2(salt, init_code_hash));
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/cast/bin/cmd/create2.rs:125]: Yeah definitely. This was migrated mainly to made the code compile, but it's completely redundant now

Also agree with SimpleCast, it could just be a CLI

File: crates/cast/bin/cmd/logs.rs [hunk 19]
```diff
@@ -208,6 +211,78 @@ fn build_filter_topics(topics: Vec<String>) -> Result<TopicFilter, eyre::Error>
...
+                    _ => sanitize_token(token),
+                };
+                sanitized.push(token)
+            }
+            Token::Array(sanitized)
+        }
+        _ => token,
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/cast/bin/cmd/logs.rs:285]: Duplicating for now—The types in this file are very tied to `ethers_providers` and we'd need the equivalent functionality to fully remove them.

File: crates/common/src/abi.rs [hunk 20]
```diff
@@ -1,50 +1,55 @@
...
-    let tokens = parse_tokens(params, true)?;
-    Ok(func.encode_input(&tokens)?)
+    Ok(func.abi_encode_input(args.as_slice())?)
 }
 
 /// Decodes the calldata of the function
 ///
 /// # Panics
 ///
 /// If the `sig` is an invalid function signature
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/common/src/abi.rs:35]: yep, outdated docs

File: crates/forge/src/runner.rs [hunk 21]
```diff
@@ -98,22 +94,18 @@ impl<'a> ContractRunner<'a> {
...
         let mut logs = Vec::new();
         let mut traces = Vec::with_capacity(self.predeploy_libs.len());
         for code in self.predeploy_libs.iter() {
-            match self.executor.deploy(
-                self.sender.to_alloy(),
-                code.0.clone().into(),
-                rU256::ZERO,
-                self.errors,
-            ) {
+            match self.executor.deploy(self.sender, code.0.clone().into(), U256::ZERO, self.errors)
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/forge/src/runner.rs:107]: Yeah this PR needs a cleanup pass—will do so tmw

File: crates/forge/bin/cmd/create.rs [hunk 22]
```diff
@@ -325,15 +324,240 @@ impl CreateArgs {
...
+};
+
+use std::{borrow::Borrow, marker::PhantomData};
+
+/// `ContractFactory` is a [`DeploymentTxFactory`] object with an
+/// [`Arc`] middleware. This type alias exists to preserve backwards
+/// compatibility with less-abstract Contracts.
+///
+/// For full usage docs, see [`DeploymentTxFactory`].
+pub type ContractFactory<M> = DeploymentTxFactory<Arc<M>, M>;
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/forge/bin/cmd/create.rs:351]: Duplicating due to ethers-providers dependency

File: crates/evm/src/executor/mod.rs (hunk 13 above)
Comment (Evalir) [reply crates/evm/src/executor/mod.rs:835]: Let's improve on this on a followup PR to not block on this—agree we should move to `alloy_sol_types`

File: crates/evm/src/fuzz/invariant/executor.rs [hunk 23]
```diff
@@ -323,26 +320,61 @@ impl<'a> InvariantExecutor<'a> {
...
+                                    (name, selectors)
+                                } else {
+                                    panic!("targetArtifactSelectors should be a tuple array")
+                                }
+                            })
+                            .collect::<Vec<_>>()
+                    } else {
+                        panic!("targetArtifactSelectors should be a tuple array")
+                    }
+                },
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/evm/src/fuzz/invariant/executor.rs:347]: Will move to `SolCall` in followup PR

File: crates/forge/bin/cmd/verify/etherscan/mod.rs [hunk 24]
```diff
@@ -524,7 +523,7 @@ mod tests {
                 &config,
             )
             .unwrap();
-        assert_eq!(client.etherscan_api_url().as_str(), "https://verifier-url.com/");
+        assert_eq!(client.etherscan_api_url().as_str(), "https://verifier-url.com/?/");
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/forge/bin/cmd/verify/etherscan/mod.rs:526]: This is due to #6079

File: crates/common/src/abi.rs [hunk 25]
```diff
@@ -246,53 +239,41 @@ impl IntoFunction for String {
 
 impl<'a> IntoFunction for &'a str {
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/common/src/abi.rs:176]: Will remove once we move to SolCall

File: crates/evm/src/decode.rs (hunk 14 above)
Comment (Evalir) [reply crates/evm/src/decode.rs]: Chatted with @DaniPopes about this, this feature is still missing from Alloy and needs to be implemented. We can leave this on abigen for now, and we'll come back to this once the feature is added, as it's only a single function

File: Cargo.toml [hunk 26]
```diff
@@ -187,18 +192,21 @@ color-eyre = "0.6"
...
+ethers-etherscan = { git = "https://github.com/gakonst/ethers-rs", rev = "9754f22d06a2defe5608c4c9b809cc361443a3dc" }
+ethers-solc = { git = "https://github.com/gakonst/ethers-rs", rev = "9754f22d06a2defe5608c4c9b809cc361443a3dc" }
+
+foundry-compilers = { git = "https://github.com/foundry-rs/compilers" }
+foundry-block-explorers = { git = "https://github.com/foundry-rs/block-explorers"}
+
+alloy-dyn-abi = { git = "https://github.com/alloy-rs/core/", branch = "dani/coerce-str"}
+alloy-primitives = { git = "https://github.com/alloy-rs/core/", branch = "dani/coerce-str"}
+alloy-json-abi = { git = "https://github.com/alloy-rs/core/", branch = "dani/coerce-str"}
+alloy-sol-types = { git = "https://github.com/alloy-rs/core/", branch = "dani/coerce-str"}
```
Comment (DaniPopes) [inline Cargo.toml:212]: merged to main

File: crates/cast/bin/cmd/estimate.rs [hunk 27]
```diff
@@ -1,20 +1,21 @@
...
 use std::str::FromStr;
 
 /// CLI arguments for `cast estimate`.
 #[derive(Debug, Parser)]
 pub struct EstimateArgs {
     /// The destination of the transaction.
-    #[clap(value_parser = NameOrAddress::from_str)]
-    to: Option<NameOrAddress>,
+    #[clap(value_parser = Address::from_str)]
+    to: Option<Address>,
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/cast/bin/cmd/estimate.rs:18]: We could make a local `NameOrAddress` since it's just `enum { Address, String }`, def not a blocker tho

File: crates/cast/bin/cmd/storage.rs [hunk 28]
```diff
@@ -33,12 +34,12 @@ const MIN_SOLC: Version = Version::new(0, 6, 5);
 #[derive(Debug, Clone, Parser)]
 pub struct StorageArgs {
     /// The contract address.
-    #[clap(value_parser = NameOrAddress::from_str)]
-    address: NameOrAddress,
+    #[clap(value_parser = Address::from_str)]
+    address: Address,
```
Comment (mattsse) [inline crates/cast/bin/cmd/storage.rs:38]: this is not equivalent because this removes ens functionality.

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by DaniPopes — APPROVED** [review]

File: crates/evm/src/fuzz/invariant/executor.rs [hunk 29]
```diff
@@ -323,26 +319,59 @@ impl<'a> InvariantExecutor<'a> {
...
+                                let name = elements[0].as_str().unwrap().to_string();
+                                let selectors = elements[1]
+                                    .as_array()
+                                    .unwrap()
+                                    .iter()
+                                    .map(|selector| {
+                                        FixedBytes::<4>::from_slice(&selector.as_fixed_bytes().unwrap().0[0..4])
+                                    })
+                                    .collect::<Vec<_>>();
+                                (name, selectors)
```
Comment (gakonst) [inline crates/evm/src/fuzz/invariant/executor.rs:339]: yeah this needs to go we cannot work with an API that is so low level IMO, these type conversions should never surface to the user

File: crates/evm/src/decode.rs [hunk 30]
```diff
@@ -185,19 +185,33 @@ pub fn decode_revert(
...
                     if err.starts_with(ERROR_PREFIX.as_slice()) {
-                        String::decode(&err[ERROR_PREFIX.len()..]).ok()
+                        DynSolType::abi_decode(&DynSolType::String, &err[ERROR_PREFIX.len()..])
+                            .map_err(|_| eyre::eyre!("Could not decode cheatcode string error"))
+                            .and_then(|v| {
+                                v.as_str()
+                                    .map(|s| s.to_owned())
+                                    .wrap_err("Bad cheatcode string decode")
+                            })
+                            .ok()
```
Comment (gakonst) [inline crates/evm/src/decode.rs:207]: this API is also more verbose than before, why doesn't `String::decode` work?

File: crates/evm/src/decode.rs [hunk 31]
```diff
@@ -129,14 +125,17 @@ pub fn decode_revert(
...
+            .map_err(|_| eyre::eyre!("Bad string decode"))
+            .and_then(|v| v.clone().as_str().map(|s| s.to_owned()).wrap_err("Bad string decode"))?
+            .to_owned()),
         // keccak(expectRevert(bytes))
         [242, 141, 206, 179] => {
             let err_data = &err[SELECTOR_LEN..];
             if err_data.len() > 64 {
-                let len = U256::from(&err_data[32..64]).as_usize();
+                let len = U256::try_from_be_slice(&err_data[32..64])
+                    .wrap_err_with(|| "Could not decode slice")?
```
Comment (gakonst) [inline crates/evm/src/decode.rs:137]: we should not be using eyre anywhere in library code, this is a good chacne to get rid of it, also if this is just 2 bytes should just deserialize it as usize in the first place

File: crates/evm/src/decode.rs [hunk 32]
```diff
@@ -1,15 +1,12 @@
...
-    prelude::U256,
-    types::Log,
-};
+use crate::executor::inspector::cheatcodes::util::MAGIC_SKIP_BYTES;
+use alloy_dyn_abi::{DynSolType, DynSolValue, JsonAbiExt};
+use alloy_json_abi::JsonAbi;
+use alloy_primitives::{B256, U256};
+use alloy_sol_types::{sol_data::String as SolString, SolType};
+use ethers::{abi::RawLog, contract::EthLogDecode, types::Log};
+use eyre::ContextCompat;
```
Comment (gakonst) [inline crates/evm/src/decode.rs:8]: remove and explicitly handle errors

File: crates/config/src/utils.rs [hunk 33]
```diff
@@ -259,6 +258,39 @@ where
...
+        }
+    }
+}
+
+impl FromStr for Numeric {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(val) = s.parse::<u128>() {
+            Ok(Numeric::U256(U256::from(val)))
```
Comment (gakonst) [inline crates/config/src/utils.rs:285]: This isn't really right, right? As you'll fail for >u128. https://docs.rs/serde_json/latest/serde_json/value/struct.Number.html

File: crates/common/src/compile.rs [hunk 34]
```diff
@@ -460,7 +460,7 @@ pub fn etherscan_project(metadata: &Metadata, target_path: impl AsRef<Path>) ->
     //     [source code]
     let paths = ProjectPathsConfig::builder()
         .sources(sources_path.clone())
-        .remappings(settings.remappings.clone())
+        // .remappings(settings.remappings.clone())
```
Comment (gakonst) [inline crates/common/src/compile.rs:463]: ?

File: crates/cli/src/utils/mod.rs [hunk 35]
```diff
@@ -134,18 +118,15 @@ where
...
 pub fn parse_ether_value(value: &str) -> Result<U256> {
     Ok(if value.starts_with("0x") {
-        U256::from_str(value)?
+        U256::from_str_radix(value, 16)?
     } else {
-        U256::from(LenientTokenizer::tokenize_uint(value)?)
+        alloy_dyn_abi::DynSolType::coerce_str(&alloy_dyn_abi::DynSolType::Uint(256), value)?
+            .as_uint()
+            .wrap_err("Could not parse ether value from string")?
+            .0
```
Comment (gakonst) [inline crates/cli/src/utils/mod.rs:126]: this got more complex now, how do we simplify it back?

**Review by gakonst — APPROVED** [review]
I'm generally supportive of moving forward with this PR but want to point out a few things:
* There's some code that was vendored over from ethers-rs to keep things going, that's fine but I would like to keep track of that code and what the plan is for it (e.g. Numeric type, unit conversions)
* There's a few APIs that seem to have gotten worse not better, e.g. the DynSolValue decoding stuff, or the req for the user to pass validate=false to the decoder
* I really dislike the `U256::from` stuff and think there _must_ be a From impl that will work with us, please investigate what the path for including From<u64> for U256 is in ruint.


Basically, I like that we're moving forward, but I think in some places we need to look at the code and say "hey actually this became worse and it looks like there's something missing from Alloy, let me fix that first".

File: crates/evm/src/fuzz/strategies/uint.rs [hunk 36]
```diff
@@ -40,33 +41,27 @@ impl UintValueTree {
...
         self.reposition()
     }
-
     fn complicate(&mut self) -> bool {
         if self.fixed || (self.hi <= self.lo) {
             return false
         }
 
-        self.lo = self.curr + 1;
+        self.lo = self.curr + U256::from(1);
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/evm/src/fuzz/strategies/uint.rs:61]: Same reason as to why there aren't From impls

File: crates/evm/src/fuzz/strategies/uint.rs [hunk 37]
```diff
@@ -101,42 +95,34 @@ impl UintStrategy {
...
-
     fn generate_edge_tree(&self, runner: &mut TestRunner) -> NewTree<Self> {
         let rng = runner.rng();
-
         // Choose if we want values around 0 or max
         let is_min = rng.gen_bool(0.5);
         let offset = U256::from(rng.gen_range(0..4));
         let max = if self.bits < 256 {
-            (U256::from(1u8) << U256::from(self.bits)) - 1
+            (U256::from(1u8).shl(self.bits)) - U256::from(1)
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/evm/src/fuzz/strategies/uint.rs:104]: Not released

File: crates/evm/src/fuzz/strategies/param.rs [hunk 38]
```diff
@@ -1,54 +1,72 @@
...
 
 /// Given a parameter type, returns a strategy for generating values for that type, given some EVM
 /// fuzz state.
 ///
 /// Works with ABI Encoder v2 tuples.
-pub fn fuzz_param_from_state(param: &ParamType, arc_state: EvmFuzzState) -> BoxedStrategy<Token> {
+pub fn fuzz_param_from_state(
+    param: &DynSolType,
+    arc_state: EvmFuzzState,
+) -> BoxedStrategy<DynSolValue> {
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/evm/src/fuzz/strategies/param.rs:69]: <https://docs.rs/alloy-dyn-abi/latest/alloy_dyn_abi/enum.DynSolValue.html#method.type_strategy>

File: crates/evm/src/fuzz/invariant/filters.rs [hunk 39]
```diff
@@ -1,21 +1,19 @@
...
 /// Contains which contracts are to be targeted or excluded on an invariant test through their
 /// artifact identifiers.
 #[derive(Default)]
 pub struct ArtifactFilters {
-    /// List of `contract_path:contract_name` which are to be targeted. If list of functions is not
-    /// empty, target only those.
-    pub targeted: BTreeMap<String, Vec<FixedBytes>>,
+    /// List of `contract_path:contract_name` along with selectors, which are to be targeted. If
+    /// list of functions is not empty, target only those.
+    pub targeted: BTreeMap<String, Vec<FixedBytes<4>>>,
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/evm/src/fuzz/invariant/filters.rs:13]: this is [`Selector`](https://docs.rs/alloy-primitives/latest/alloy_primitives/aliases/type.Selector.html)

File: crates/evm/src/fuzz/invariant/executor.rs [hunk 40]
```diff
@@ -522,7 +585,32 @@ impl<'a> InvariantExecutor<'a> {
...
+                            } else {
+                                panic!("targetSelectors should be a tuple array2")
+                            }
+                        })
+                        .collect::<Vec<_>>()
+                } else {
+                    panic!("targetSelectors should be a tuple array")
+                }
+            });
```
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/evm/src/fuzz/invariant/executor.rs:613]: https://github.com/foundry-rs/foundry/pull/5986#discussion_r1370226176

File: crates/evm/src/fuzz/strategies/invariants.rs [hunk 41]
```diff
@@ -100,76 +92,69 @@ fn select_random_sender(
...
         ),
     ])
     // Too many exclusions can slow down testing.
     .prop_filter("senders not allowed", move |addr| !senders_ref.excluded.contains(addr))
     .boxed();
-
     if !senders.targeted.is_empty() {
         any::<prop::sample::Selector>()
             .prop_map(move |selector| *selector.select(&*senders.targeted))
+            .prop_map(|s| s)
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/evm/src/fuzz/strategies/invariants.rs:112]: noop, removed

File: crates/cli/src/utils/mod.rs (hunk 35 above)
Comment (Evalir) [reply crates/cli/src/utils/mod.rs:126]: A move to `SolCall` should simplify this—will be done on a followup

File: crates/evm/src/fuzz/strategies/calldata.rs [hunk 42]
```diff
@@ -1,18 +1,26 @@
...
 /// for that function's input types.
 pub fn fuzz_calldata(func: Function) -> BoxedStrategy<Bytes> {
     // We need to compose all the strategies generated for each parameter in all
     // possible combinations
-    let strats = func.inputs.iter().map(|input| fuzz_param(&input.kind)).collect::<Vec<_>>();
+    let strats = func
+        .inputs
+        .iter()
+        .map(|input| fuzz_param(&DynSolType::from_str(&input.selector_type()).unwrap()))
+        .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/evm/src/fuzz/strategies/calldata.rs:18]: Before, `fuzz_param` used `ParamType` which was analogous to `DynSolType` but limited to function params. Now we just have the type string information which we need to convert to a `DynSolType`. I think it'd be nice to be able to convert from a function `Param` to a `DynSolType` directly

File: crates/evm/src/decode.rs (hunk 30 above)
Comment (Evalir) [reply crates/evm/src/decode.rs:207]: `String::decode` came from the `AbiDecode` trait—we don't have an analogous trait in Alloy. We could probably get one? cc @DaniPopes

File: Cargo.toml [hunk 43]
```diff
@@ -187,18 +192,21 @@ color-eyre = "0.6"
...
+ethers-etherscan = { git = "https://github.com/gakonst/ethers-rs", rev = "9754f22d06a2defe5608c4c9b809cc361443a3dc" }
+ethers-solc = { git = "https://github.com/gakonst/ethers-rs", rev = "9754f22d06a2defe5608c4c9b809cc361443a3dc" }
+
+foundry-compilers = { git = "https://github.com/foundry-rs/compilers" }
+foundry-block-explorers = { git = "https://github.com/foundry-rs/block-explorers"}
+
+alloy-dyn-abi = { git = "https://github.com/alloy-rs/core/" }
+alloy-primitives = { git = "https://github.com/alloy-rs/core/" }
+alloy-json-abi = { git = "https://github.com/alloy-rs/core/" }
+alloy-sol-types = { git = "https://github.com/alloy-rs/core/" }
```
[the remark this answers is not in the document]
Comment (Evalir) [reply Cargo.toml:212]: The main offender is `foundry-block-explorers` needing the chain enum from ethers. Once we have `alloy-rs/chains` we can publish both crates and remove the patches. For alloy, we need a newer release with all the new features we're using off of master.

File: crates/evm/src/trace/decoder.rs [hunk 44]
```diff
@@ -143,37 +138,37 @@ impl CallTraceDecoder {
...
-                0x02: sha256(data: Bytes) -> (hash: FixedBytes(32)),
-                0x03: ripemd(data: Bytes) -> (hash: FixedBytes(32)),
-                0x04: identity(data: Bytes) -> (data: Bytes),
-                0x05: modexp(Bsize: Uint(256), Esize: Uint(256), Msize: Uint(256), BEM: Bytes) -> (value: Bytes),
-                0x06: ecadd(x1: Uint(256), y1: Uint(256), x2: Uint(256), y2: Uint(256)) -> (x: Uint(256), y: Uint(256)),
-                0x07: ecmul(x1: Uint(256), y1: Uint(256), s: Uint(256)) -> (x: Uint(256), y: Uint(256)),
-                0x08: ecpairing(x1: Uint(256), y1: Uint(256), x2: Uint(256), y2: Uint(256), x3: Uint(256), y3: Uint(256)) -> (success: Uint(256)),
-                0x09: blake2f(rounds: Uint(4), h: FixedBytes(64), m: FixedBytes(128), t: FixedBytes(16), f: FixedBytes(1)) -> (h: FixedBytes(64)),
+                0x01: ecrecover(hash: format!("bytes32"), v: format!("uint256"), r: format!("uint256"), s: format!("uint256")) -> (publicAddress: format!("address")),
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/evm/src/trace/decoder.rs:141]: Tracking on https://github.com/foundry-rs/foundry/issues/6102

**Comment by Evalir** [comment]
Merging—CI is flaky due to arbitrum rpc limiting (happening on other PRs as well)
